src/storage/txn/store.rs

```diff
@@ -208,33 +208,79 @@ impl<'a> SnapshotStore<'a> {
 
     pub fn scanner(&self) -> Result<StoreScanner> {
         let cursor = try!(self.snapshot.iter());
-        let txn = MvccSnapshot::new(self.snapshot, self.start_ts);
         Ok(StoreScanner {
             cursor: cursor,
-            txn: txn,
+            start_ts: self.start_ts,
         })
     }
 }
 
 pub struct StoreScanner<'a> {
     cursor: Box<Cursor + 'a>,
-    txn: MvccSnapshot<'a>,
+    start_ts: u64,
 }
 
 impl<'a> StoreScanner<'a> {
+    pub fn seek(&mut self, mut key: Key) -> Result<Option<(Key, u64)>> {
+        loop {
+            if !try!(self.cursor.seek(&key)) {
+                return Ok(None);
+            }
+            key = try!(Key::from_encoded(self.cursor.key().to_vec()).truncate_ts());
+            let cursor = self.cursor.as_mut();
+            let mut txn = MvccCursor::new(cursor, self.start_ts);
+            if let Some(ts) = try!(txn.get_version(&key)) {
+                return Ok(Some((key, ts)));
+            }
+            key = key.append_ts(u64::max_value());
+        }
+    }
+
+    pub fn reverse_seek(&mut self, mut key: Key) -> Result<Option<(Key, u64)>> {
+        loop {
+            if !try!(self.cursor.reverse_seek(&key)) {
+                return Ok(None);
+            }
+            key = try!(Key::from_encoded(self.cursor.key().to_vec()).truncate_ts());
+            let cursor = self.cursor.as_mut();
+            let mut txn = MvccCursor::new(cursor, self.start_ts);
+            if let Some(ts) = try!(txn.get_version(&key)) {
+                return Ok(Some((key, ts)));
+            }
+        }
+    }
+
+    #[inline]
+    fn handle_mvcc_err(e: MvccError, result: &mut Vec<Result<KvPair>>) -> Result<Key> {
+        let key = if let MvccError::KeyIsLocked { key: ref k, .. } = e {
+            Some(Key::from_raw(k))
+        } else {
+            None
+        };
+        match key {
+            Some(k) => {
+                result.push(Err(e.into()));
+                Ok(k)
+            }
+            None => Err(e.into()),
+        }
+    }
+
     pub fn scan(&mut self, mut key: Key, limit: usize) -> Result<Vec<Result<KvPair>>> {
         let mut results = vec![];
         while results.len() < limit {
-            if !try!(self.cursor.seek(&key)) {
-                break;
+            match self.seek(key) {
+                Ok(Some((k, ts))) => {
+                    // should we panic if None?
```

Comment:
My mistake, it's a normal case to get a `None` here. Remove this comment please.